includes/admin/class-amp-story-templates.php

```diff
@@ -0,0 +1,197 @@
+	 * Init.
+	 */
+	public function init() {
+		if ( ! post_type_exists( 'amp_story' ) ) {
+			return;
+		}
+
+		add_filter( 'rest_wp_block_query', array( $this, 'filter_rest_wp_block_query' ), 10, 2 );
+
+		$this->register_taxonomy();
+		$this->maybe_import_story_templates();
+	}
+
+	/**
+	 * Import story templates if they haven't been imported previously.
+	 */
+	public function maybe_import_story_templates() {
+		if ( self::STORY_TEMPLATES_VERSION === AMP_Options_Manager::get_option( 'story_templates_version' ) ) {
+			return;
+		}
+		$this->import_story_templates();
+		// @todo This is commented out for testing purposes until the PR gets ready.
+		// AMP_Options_Manager::update_option( 'story_templates_version', self::STORY_TEMPLATES_VERSION );
+	}
+
+	/**
+	 * Import logic for story templates.
+	 */
+	public function import_story_templates() {
+		$templates = self::get_story_templates();
+		foreach ( $templates as $template ) {
+			$existing_template = $this->template_exists( $template['title'], $template['content'] );
+			if ( $existing_template > 0 ) {
+				continue;
+			}
+			$post_id = wp_insert_post(
+				array(
+					'post_title'   => $template['title'],
+					'post_type'    => 'wp_block',
+					'post_status'  => 'publish',
+					'post_content' => $template['content'],
+				)
+			);
+			wp_set_object_terms( $post_id, self::TEMPLATES_TERM, self::TEMPLATES_TAXONOMY );
+		}
+	}
+
+	/**
+	 * Get the static list of templates, including temlate name and content.
+	 *
+	 * @return array Story templates.
+	 */
+	public static function get_story_templates() {
+		return array(
+			array(
+				'title'   => __( 'Default Template 1', 'amp' ),
+				'content' => '<!-- wp:amp/amp-story-page {"backgroundColor":"#313131","autoAdvanceAfterDuration":0} -->
+<amp-story-page style="background-color:#313131" id="e7ae309f-d9f1-480a-a35e-7a495c2e09f7" class="wp-block-amp-amp-story-page"><amp-story-grid-layer template="vertical"><!-- wp:amp/amp-story-text {"tagName":"h1","autoFontSize":28,"ampFontFamily":"Garamond","textColor":"very-light-gray","positionTop":27,"positionLeft":13} -->
+<h1 style="font-size:28px;width:76%;height:9%;position:absolute;top:27%;left:13%" class="has-text-color has-very-light-gray-color" id="ddce50c3-5776-44c5-9d9b-0d6d4b12d88e" data-font-family="Garamond"><amp-fit-text layout="fill" class="amp-text-content">Hello, Templates!</amp-fit-text></h1>
+<!-- /wp:amp/amp-story-text --></amp-story-grid-layer></amp-story-page>
+<!-- /wp:amp/amp-story-page -->',
+			),
+		);
+	}
+
+	/**
+	 * Determine if a post exists based on title, content, only receives a published post.
+	 *
+	 * Note that this is almost a copy of the default `post_exists` function except for the post status.
+	 *
+	 * @global wpdb $wpdb WordPress database abstraction object.
+	 *
+	 * @param string $title Post title.
+	 * @param string $content Post content.
+	 * @return int Post ID if post exists, 0 otherwise.
+	 */
+	private function template_exists( $title, $content ) {
```

Comment:
Instead of checking if a post exists with a given `$title` and `$content`. What about just using a slug? The `get_story_templates` method could include a `name` in addition to a `title` and `content`, and the name could be something like `amp-story-template-1` (or even a UUID). Then to check if the template exists, just check if a `wp_block` exists with that `post_name` (slug).

Reply:
Hmm, but this way re-importing (a future feature) wouldn't work as expected after editing the templates since the slugs would stay the same. The idea was that if the default template has been changed then re-importing would create new templates.

Perhaps we should check the content only instead?

Reply:
OK, good point. 

Note also that there is a `post_exists()` which could potentially be re-used?

Otherwise, the slug alone could be checked, and when a template is inserted by the AMP plugin here, a postmeta flag could be added to indicate it is `original` (or unmodified). Then, there could be a `save_post_amp_story` action which is used to remove this postmeta whenever the post is subsequently edited (by a means other than the AMP plugin here). When that postmeta is absent, then we would know that the template should _not_ be overridden since the user modified it. Otherwise, if the `original` flag _is_ present, then we could freely update the template. This would ensure that new versions of the template could be saved into the database such as when the templates themselves evolve.

Reply:
I think that `post_exists()` doesn't include the check for `publish` status, it would give a positive even for trashed templates.

> This would ensure that new versions of the template could be saved into the database such as when the templates themselves evolve.

That's an interesting idea. The `TEMPLATE_VERSION` constant was added for that reason -- to track if the templates have evolved meanwhile to automatically start the import (if the version is changed). However, it's true that it would just import everything instead of updating the previous ones.

One concern with updating the previous templates is that the user might actually be using them and not want an update. Thoughts?

Reply:
> I think that `post_exists()` doesn't include the check for `publish` status, it would give a positive even for trashed templates.

OK, but then you could check for the `status` after it is returned.

> One concern with updating the previous templates is that the user might actually be using them and not want an update. Thoughts?

I guess it depends on how we expect the templates to evolve over time. Would they necessarily change much? Should we just assume the overall look and feel of the template is pretty well settled and we'd make incremental changes? In other words, they could be treated like core themes: the design does not change, but the code does evolve a bit over time.

Reply:
> OK, but then you could check for the status after it is returned.

`post_exists()` returns only one post's ID (the first match). If we have one Trashed template and another published and the first match would be the one that has been trashed then we'd still end up importing a duplicate template. The method in the PR is actually based on `post_exists` since that was the initial thought to use, unfortunately, it doesn't allow defining the status to check against, would be useful here.

> Would they necessarily change much? Should we just assume the overall look and feel of the template is pretty well settled and we'd make incremental changes?

Probably it wouldn't change a lot, true.

On checking against the slug only: if we would have predefined names (slugs) and the template would be edited by the user, and then reimport would be run. This would mean that a new template would be created with a new slug (not the predefined one since that would already exist), would we then have to store the new slugs for being able to check later if the reimported template with a new slug has the `original` flag? Probably we should then always only check the latest "new" slug of that specific template when reimporting? Thoughts?

We could use the `original` flag in any case.

Reply:
OK, cool on not using `post_exists()`.

If a user modified the original template, then how about juts skipping it? If the user really wants to get the original template back, then they could delete the modified template and then run the import again?

Reply:
> If a user modified the original template, then how about just skipping it?

What if the user wants both, the original and the modified? 😄 

>  If the user really wants to get the original template back, then they could delete the modified template and then run the import again?

In this case, we should maybe display a message when running the import to indicate that some of the templates already existed and weren't imported? This could hint that deleting the previous templates would cause re-importing the templates, otherwise, it wouldn't be clear why the templates aren't being imported.

Reply:
If that is easy I suppose. But doing whatever is most straightforward to start with is probably best. Then we can iterate.